crates/ra_hir/src/ty.rs

```diff
@@ -973,14 +1143,24 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
                 ret_ty
             }
             Expr::Match { expr, arms } => {
-                let _ty = self.infer_expr(*expr, &Expectation::none());
-                for arm in arms {
-                    // TODO type the bindings in pats
+                let mut expected = expected.clone();
+                let input_ty = self.infer_expr(*expr, &Expectation::none());
+                let pat_expectation = Expectation::has_type(input_ty);
+
+                for MatchArm {
```

Comment:
tbh I think this would be more readable without destructuring here... `arm.expr` isn't even longer than `arm_expr` ;)